[QAE-456] Simple bolus calculator element updates

Foscottl-TP opened the pull request:
Added multiple elements in support of the Simple Bolus Calculator test case automation

Foscottl-TP committed:
[QAE-456] New objects, CGM pill read func, and extension

Changed bolusEntryTextField so it will support SImple Bolus Calulator and Bolus screens.
Both Current Glucose and Bolus fields use the same identifier so rather than the default value determining the correct field, a supporting extension in Base Screen was added to help distinguish the the two.
This can be used for parsing similar situations in future test cases.

Foscottl-TP committed:
[QAE-456] New objects, CGM pill read func, and extension

Changed bolusEntryTextField so it will support SImple Bolus Calulator and Bolus screens.
Both Current Glucose and Bolus fields use the same identifier so rather than the default value determining the correct field, a supporting extension in Base Screen was added to help distinguish the the two.
This can be used for parsing similar situations in future test cases.

Foscottl-TP committed:
Rebase fix

HomeScreen push was updated before updating with current build. This is to undo the previous commit.

Foscottl-TP committed:
Testing Label Change for application in Simple Bolus Calculator

Changing the Label names to see if this will work in application rather than using the first/last match method

Foscottl-TP committed:
Text field label test V2

Testing again if the bolus and current glucose field can be separately defined other than dissmissablekeyboardtextfield.

Foscottl-TP committed:
Update element query text

Updated text field query for Current Glucose and Bolus elements

Foscottl-TP committed:
More experiments with pulling dismissibleKeyboardText Field elements

Yet another attempt to pull duplicate identifier fields

Foscottl-TP committed:
FinalUpdate for CurrentGlucose/Bolus Labels

Buckled and just added accesibilityIdentifier to each element...

Foscottl-TP committed:
Read Bolus Field action

Added Action to  get the bolus field value. Needed for verifying if the value is reset when the current glucose field is edited.

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -12,39 +12,63 @@ public final class BolusScreen: BaseScreen {
     private var bolusCancelButton: XCUIElement { app.navigationBars.buttons["Cancel"] }
     private var simpleBolusCalculatorTitle: XCUIElement { app.navigationBars.staticTexts["Simple Bolus Calculator"] }
     private var bolusActionButton: XCUIElement { app.buttons["button_bolusAction"] }
+    private var saveAndDeliverButton: XCUIElement { app.buttons["Save and Deliver"] }
```
No need to add this page object. It is already implemented as 'bolusActionButton'. Reason why it is named like this is, that Bolus screen can contains different values of the button but we can interact with it using the same accessibilityIdentifier. E.g. if we want to know that button has some specific text we can create function like 'func getBolusActionButtonText() { bolusActionButton.getLabelSafe() }'. Possible values of the button on the Bolus screen are displayed within the attached screenshot. 

So instead of using this page object you can use 'bolusActionButton' and it's method 'tapBolusActionButton()'.

![image](https://github.com/user-attachments/assets/8cf90fa5-6d11-429b-a7f7-327f9032b5fd)

Foscottl-TP replied:
I believe button_bolusAction indentifier was applied to the Bolus screen but not the Simple Bolus Calculator previously. I will update that so both screen act consistently and can use the element call in the same way.

Petr-Zywczok replied:
You are right. Thanks.

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -12,39 +12,63 @@ public final class BolusScreen: BaseScreen {
 
     private var bolusTitleText: XCUIElement { app.navigationBars.staticTexts["Bolus"] }
-    private var bolusEntryTextField: XCUIElement { app.textFields["dismissibleKeyboardTextField"] }
+    private var currentGlucoseEntryTextField: XCUIElement { app.textFields["Current Glucose Field"]}
```
Please add accessibilityIdentifier in format 'elementType_ElementID'. In this case 'textField_CurrentGlucose'.

Same comment applies to other elements: bolusEntryTextField, saveAndDeliverButton, glucoseEntryRangeWarning

Foscottl-TP replied:
Updated naming convention for textFields: Current Glucose and Bolus. Save and Deliver was removed in favor of  adding the button_BolusAction identifier to all of the permutations of the button (like the Bolus screen). glucoseEntryRangeWarning was updated to glucoseEntryOutOfRangeWarning for clarity and the identifier was updated as described below

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -12,39 +12,63 @@ public final class BolusScreen: BaseScreen {
     
+   
+    // Warnings
+    private var glucoseEntryRangeWarning: XCUIElement { app.staticTexts["Glucose Entry Out of Range"] }
```
This element is already identified by accessibilityIdentifier `text_guardrailWarning`.

When you let print screen tree in debug via command `po app.descendants(matching: .any)` you can see there is already set an identifier `text_guardrailWarning`. This one is used in TherapySettingsScreen to define element `guardrailWarningText` which has functions to check if it exists and get the label of this element. So you can check if it displays "Glucose Entry Out of Range" text or any other that is supported for this element.

![image](https://github.com/user-attachments/assets/2b9f4913-bce8-4641-a521-cc692af8f774)

Foscottl-TP replied:
Changed to the text_guardrailWarning as suggested. Thanks for the tip for finding identifiers unseen in the accessibility inspector! :) Super useful!

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BaseScreen.swift:
```diff
@@ -7,6 +7,10 @@
 import XCTest
 
+extension XCUIElementQuery {
+    var lastMatch : XCUIElement {return self.element(boundBy: self.count - 1)}
```
Please fix code formatting to use one space after `{` and before `}`
```suggestion
    var lastMatch : XCUIElement { return self.element(boundBy: self.count - 1) }
```

Foscottl-TP replied:
Formatting fixed

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -12,39 +12,63 @@ public final class BolusScreen: BaseScreen {
             .typeText(String(repeating: XCUIKeyboardKey.delete.rawValue, count: currentTextLength))
     }
     
+   
+    
```
Remove unnecessary new lines.

Foscottl-TP replied:
Removed new lines

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -12,39 +12,63 @@ public final class BolusScreen: BaseScreen {
+    private var glucoseEntryRangeWarning: XCUIElement { app.staticTexts["Glucose Entry Out of Range"] }
+    
+    
+    
+    
+    
```
Remove unnecessary new lines.

Foscottl-TP replied:
Removed new lines

Foscottl-TP committed:
Update nameing conventions

Updated to corrent naming conventions

Foscottl-TP committed:
Rebase error fix

Foscottl-TP committed:
Spacing & Guardrail Warning

Fixed the spacing formatting issue on a Base Screen function and changed the label of the glucose EntryOutOfRangeWarning to match an identifier already present.

Foscottl-TP committed:
Removed access new lines

What the summary says

Foscottl-TP committed:
NameUpdate

Fixed incorrect field name

Foscottl-TP committed:
Added getCurrentGlucoseFieldValue

Used to support the optional in verification: "Then x field displays value x" (New Bolus steps addition)

Foscottl-TP committed:
CapitalizationRevert

Reverted Capitalization and added Carbohydrates

Foscottl-TP committed:
Removal of Glucose guardrail warning

Removed element and verification because it is no longer used due to suggested change in LoopSupport

Foscottl-TP committed:
Removal of Glucose guardrail warning

Removed element and verification because it is no longer used due to suggested change in LoopSupport

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/HomeScreen.swift:
```diff
@@ -52,6 +53,11 @@ public final class HomeScreen: BaseScreen {
+
+
+    public func getPumpPillValue() -> String { hudPumpPill.getValueSafe() }
+    public func getHudGlucosePill() -> String { hudGlucosePill.getValueSafe()}
```
```suggestion
    public func getHudGlucosePill() -> String { hudGlucosePill.getValueSafe() }
```

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -41,6 +52,7 @@ public final class BolusScreen: BaseScreen {
     }
     
     public func setBolusEntryTextField(value: String) { bolusEntryTextField.typeText(value) }
+    public func setCurrentGlucoseEntryTextField(value: String) { currentGlucoseEntryTextField.typeText(value)}
```
```suggestion
    public func setCurrentGlucoseEntryTextField(value: String) { currentGlucoseEntryTextField.typeText(value) }
```

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -25,9 +27,18 @@ public final class BolusScreen: BaseScreen {
     public var getBolusActionButtonLabel: String { bolusActionButton.getLableSafe() }
     
     public func tapCancelBolusButton() { bolusCancelButton.safeTap() }
+    public func tapCurrentGlucoseEntryTextField() {currentGlucoseEntryTextField.safeTap()}
```
```suggestion
    public func tapCurrentGlucoseEntryTextField() { currentGlucoseEntryTextField.safeTap() }
```

Petr-Zywczok commented on Sources/LoopUITestingKit/Screens/BolusScreen.swift:
```diff
@@ -12,7 +12,9 @@ public final class BolusScreen: BaseScreen {
 
     private var bolusTitleText: XCUIElement { app.navigationBars.staticTexts["Bolus"] }
-    private var bolusEntryTextField: XCUIElement { app.textFields["dismissibleKeyboardTextField"] }
+    private var currentGlucoseEntryTextField: XCUIElement { app.textFields["textField_CurrentGlucose"]}
+    private var bolusEntryTextField: XCUIElement { app.textFields["textField_Bolus"]}
+    private var carbohydratesTextField: XCUIElement { app.textFields["textField_Carbohydrates"]}
```
```suggestion
    private var currentGlucoseEntryTextField: XCUIElement { app.textFields["textField_CurrentGlucose"] }
    private var bolusEntryTextField: XCUIElement { app.textFields["textField_Bolus"] }
    private var carbohydratesTextField: XCUIElement { app.textFields["textField_Carbohydrates"] }
```

Petr-Zywczok reviewed (approved):
There are some formatting comments, the rest looks good.

Foscottl-TP committed:
Merge branch 'main' into scott/QAE-456-SimpleBolusCalculator-MinandMaxGlucos

Foscottl-TP committed:
Merge branch 'scott/QAE-456-SimpleBolusCalculator-MinandMaxGlucos' of https://github.com/tidepool-org/LoopUITestingKit into scott/QAE-456-SimpleBolusCalculator-MinandMaxGlucos
